### PR TITLE
[Example] Progressive Sample Selection (PSS).

### PR DESCRIPTION
**Pull request by tianyuecao**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [ ] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by mufeili** [comment]
Could you or @sneakerkg @hetong007 provide more context for this example before we proceed? Is this your paper?

**Comment by tianyuecao** [comment]
Hi Mufei,


Yes, it's my ECCV22 paper during internship and now wanna release the code to dgl lib. For details, you can ask tianjun or tong.


Thanks,
Tianyue



---Original---
From: "Mufei ***@***.***&gt;
Date: Mon, Jul 18, 2022 17:57 PM
To: ***@***.***&gt;;
Cc: "Tianyue ***@***.******@***.***&gt;;
Subject: Re: [dmlc/dgl] [Example] Progressive Sample Selection (PSS). (PR #4263)




 
Could you or @sneakerkg @hetong007 provide more context for this example before we proceed? Is this your paper?

File: examples/pytorch/PSS/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# PSS
+
+Code for the ECCV '22 submission "PSS: Progressive Sample Selection for Open-World Visual Representation Learning".
```
Comment (sneakerkg) [inline examples/pytorch/PSS/README.md:3]: Not a submission anymore.

File: examples/pytorch/PSS/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+## Dependencies
+
+We use python 3.7. The CUDA version needs to be 10.2. Besides DGL (>=0.5.2), we depend on several packages. To install dependencies using conda:
```
Comment (sneakerkg) [inline examples/pytorch/PSS/README.md:7]: Can you be more precise about DGL version? Recently there's a DGL upgrade that changes dataloader interface.

File: examples/pytorch/PSS/README.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+```
+
+## Data
+
+We use the iNaturalist 2018 dataset. 
+- download link: https://www.kaggle.com/c/inaturalist-2018/data;
+- annotations are in `Smooth_AP/data/Inaturalist`.
+
+## Pretrained Models
+We support ResNet-50 and ViT-Base/16 backbones. For ViT-Base/16 backbone, download [DINO pretrained model](https://dl.fbaipublicfiles.com/dino/dino_vitbase16_pretrain/dino_vitbase16_pretrain.pth) and use `--pretrained_weights` in training Smooth_AP.
```
Comment (sneakerkg) [inline examples/pytorch/PSS/README.md:36]: @mufeili Do we have any policy on putting the link for third party models?

File: examples/pytorch/PSS/README.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+bash scripts/train_inat.sh # train iter 0 hilander model
+bash scripts/test_inat.sh # get iter 0 selected samples and pseudo labels
+
+# iteration 1
+cd ../Smooth_AP
+bash scripts/finetune_1head_inat.sh # train iter 1 feature extractor
+bash scripts/get-features-inat.sh # iter 1 generate features for hilander
+cd ../hilander
+bash scripts/train_inat.sh # train iter 1 hilander model
+bash scripts/test_inat.sh # get iter 1 selected samples and pseudo labels
```
Comment (sneakerkg) [inline examples/pytorch/PSS/README.md:55]: I remember we go for 4 iters, why there're only two? And maybe we want to push all this into a single .sh file.

**Comment by mufeili** [comment]
@jermainewang It looks like this is the implementation of a paper from the lab. See if you have further questions/concerns.

File: examples/pytorch/PSS/README.md (hunk 3 above)
Comment (mufeili) [reply examples/pytorch/PSS/README.md:36]: We recommend following the practice of [the GCN example](https://github.com/dmlc/dgl/tree/master/examples/pytorch/gcn) for the README file and the script.

File: examples/pytorch/hilander/PSS/test.sh [hunk 5]
```diff
@@ -0,0 +1,9 @@
+python Smooth_AP/src/evaluate_model.py \
+--dataset Inaturalist \
+--bs 384 \
+--source_path ~/code/Smooth_AP/data/ --embed_dim 128 \
+--resume "INATURALIST_RESNET50_2021-9-2-4-21-21/checkpoint.pth.tar" \
```
Comment (sneakerkg) [inline examples/pytorch/hilander/PSS/test.sh:5]: Users won't have the INATURALIST_RESNET50_2021-9-2-4-21-21 , maybe we can put a place holder and let user type in the exact path.

**Comment by sneakerkg** [comment]
Trying to see if we can get rid of put __init__ under the example folder. Also, need to know the exact changes in smoothAP code to see if we can use a link instead of putting all the code here.

**Review by sneakerkg — CHANGES_REQUESTED** [review]

**Comment by jermainewang** [comment]
Please delete `examples/pytorch/__init__.py`

File: examples/pytorch/hilander/PSS/Smooth_AP/src/auxiliaries.py [hunk 6]
```diff
@@ -0,0 +1,998 @@
...
+    kmeans.n_iter, kmeans.min_points_per_centroid, kmeans.max_points_per_centroid = 20,5,1000000000
+    kmeans.train(features)
+    _, cluster_assignments = kmeans.index.search(features,1)
+    return cluster_assignments
+
+
+
+"""============================================================================================================="""
+def save_graph(opt, model):
+    """
```
Comment (Jim61C) [inline examples/pytorch/hilander/PSS/Smooth_AP/src/auxiliaries.py:717]: we don't need save_graph support for viz, can we take this out?

File: examples/pytorch/hilander/PSS/Smooth_AP/src/auxiliaries.py [hunk 7]
```diff
@@ -0,0 +1,998 @@
...
+        metric_dict['val'] = ['Epochs','Time','NMI', 'F1']
+        metric_dict['val'] += ['Recall @ {}'.format(k) for k in k_vals]
+
+    return metric_dict
+
+
+
+"""================================================================================================="""
+def run_kmeans(features, n_cluster):
+    """
```
Comment (Jim61C) [inline examples/pytorch/hilander/PSS/Smooth_AP/src/auxiliaries.py:697]: Ditto, don't think we need kmeans in PSS training

File: examples/pytorch/hilander/PSS/Smooth_AP/src/auxiliaries.py [hunk 8]
```diff
@@ -0,0 +1,998 @@
...
+    if not os.path.exists(checkfolder):
+        os.makedirs(checkfolder)
+    opt.save_path = checkfolder
+
+    #Store training parameters as text and pickle in said folder.
+    with open(opt.save_path+'/Parameter_Info.txt','w') as f:
+        f.write(gimme_save_string(opt))
+    pkl.dump(opt,open(opt.save_path+"/hypa.pkl","wb"))
+
+import pdb
```
Comment (Jim61C) [inline examples/pytorch/hilander/PSS/Smooth_AP/src/auxiliaries.py:566]: remove import pdb.

File: examples/pytorch/hilander/PSS/Smooth_AP/src/auxiliaries.py [hunk 9]
```diff
@@ -0,0 +1,998 @@
...
+            Nothing!
+        """
+        with open(self.save_path, "a") as csv_file:
+            writer = csv.writer(csv_file, delimiter=',')
+            writer.writerow(inputs)
+
+
+
+################## PLOT SUMMARY IMAGE #####################
+class InfoPlotter():
```
Comment (Jim61C) [inline examples/pytorch/hilander/PSS/Smooth_AP/src/auxiliaries.py:461]: Ditto, don't think this is necessary for the minimal POC training.

File: examples/pytorch/hilander/PSS/Smooth_AP/src/losses.py [hunk 10]
```diff
@@ -0,0 +1,357 @@
+# repo originally forked from https://github.com/Confusezius/Deep-Metric-Learning-Baselines
+
+###################### LIBRARIES #################################################
```
Comment (Jim61C) [inline examples/pytorch/hilander/PSS/Smooth_AP/src/losses.py:3]: what's the difference of the loss.py as compared to Smooth_AP_loss.py? or are they duplicates of each other?

File: examples/pytorch/hilander/PSS/Smooth_AP/src/evaluate.py [hunk 11]
```diff
@@ -0,0 +1,377 @@
+# repo originally forked from https://github.com/Confusezius/Deep-Metric-Learning-Baselines
+
```
Comment (Jim61C) [inline examples/pytorch/hilander/PSS/Smooth_AP/src/evaluate.py:2]: I see that neither main.py and finetune_1head.py uses auxiliaries_nofaiss.py, rather, both use auxiliaries.py, can we remove this file?

File: examples/pytorch/hilander/PSS/Smooth_AP/src/losses.py (hunk 10 above)
Comment (hetong007) [reply examples/pytorch/hilander/PSS/Smooth_AP/src/losses.py:3]: Would be highly appreciated if you could send out reviews in a single batch, otherwise my inbox is bombed :)

File: examples/pytorch/hilander/PSS/Smooth_AP/src/datasets.py [hunk 12]
```diff
@@ -0,0 +1,1295 @@
...
+    eval_dataset        = BaseTripletDataset(train_image_dict, opt, is_validation=True)
+
+    #train_dataset.conversion       = conversion
+    #val_dataset.conversion         = conversion
+    #eval_dataset.conversion        = conversion
+    
+    return {'training':train_dataset, 'testing':val_dataset, 'evaluation':val_dataset}
+
+
+################## BASIC PYTORCH DATASET USED FOR ALL DATASETS ##################################
```
Comment (Jim61C) [inline examples/pytorch/hilander/PSS/Smooth_AP/src/datasets.py:616]: can we do imports from original SmoothAP code base for these dataset declarations?

File: examples/pytorch/hilander/PSS/Smooth_AP/src/datasets.py [hunk 13]
```diff
@@ -0,0 +1,1295 @@
...
+            cls = batch_item[0]
+        img = Image.open(str(batch_item[1]))
+        return cls, self.transform(self.ensure_3dim(img))
+
+
+    def __len__(self):
+        return len(self.dataset)
+
+
+class TrainDatasetsmoothapSoft(Dataset):
```
Comment (Jim61C) [inline examples/pytorch/hilander/PSS/Smooth_AP/src/datasets.py:1120]: Soft labels are not used in the end, can we remove?

File: examples/pytorch/hilander/PSS/train.sh [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+--class_num 948 --loss smoothap \
+--trainset lin_train_set1.txt \
+--all_trainset train_set1.txt \
+--testset Inaturalist_test_set1.txt \
+--linsize 29011
+
+# iter 0 (supervised baseline) - train hi-lander
+python train_subg_inat.py \
+--data_path "/home/ubuntu/code/dgl/examples/pytorch/hilander/PSS/data/Inaturalist/T_train_iter0_smoothap_inat_features.pkl" \
+--model_filename '/home/ubuntu/code/dgl/examples/pytorch/hilander/PSS/hilander_checkpoint/inat_l_smoothap_iter0.pth' \
```
Comment (Jim61C) [inline examples/pytorch/hilander/PSS/train.sh:29]: change to relative paths?

**Review by Jim61C — CHANGES_REQUESTED** [review]
Please refer to the individual comments for details. Overall, there still seem to be many functions that are not used for the PSS method in the paper, such as softlabels etc. Also, functions such as plotting, model graph logging are not necessary. Several datasets seem to be directly from the original SmoothAP repo, suggestion is to change to directly import from the official SmoothAP repo if possible.

**Comment by Jim61C** [comment]
Sorry that just saw this, sure, I will avoid putting down individual
comments.

Best,
Yifan

File: examples/pytorch/hilander/PSS/Smooth_AP/src/evaluate.py (hunk 11 above)
Comment (tianyuecao) [reply examples/pytorch/hilander/PSS/Smooth_AP/src/evaluate.py:2]: Yes. auxiliaries_nofaiss.py removed.

File: examples/pytorch/hilander/PSS/README.md [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+|    |- train_set1.txt
+|    |- uin_train_set1.txt
+|    |- uout_train_set1.txt
+|    |- in_train_set1.txt
+|    |- Inaturalist_test_set1.txt
+|-...
+```
+
+## Pretrained Models
+We support ResNet-50 and ViT-Base/16 backbones. For ViT-Base/16 backbone, download [DINO pretrained model](https://dl.fbaipublicfiles.com/dino/dino_vitbase16_pretrain/dino_vitbase16_pretrain.pth) and use `--pretrained_weights` in training Smooth_AP.
```
Comment (sneakerkg) [inline examples/pytorch/hilander/PSS/README.md:59]: Delete DINO, and where is the ResNet-50 link?

File: examples/pytorch/hilander/PSS/README.md [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+## Training
+We provide training scripts for feature extractor Smooth_AP and clustering model Hi-LANDER.
```
Comment (sneakerkg) [inline examples/pytorch/hilander/PSS/README.md:62]: Update this. Now it should be a single sh file.

File: examples/pytorch/hilander/PSS/Smooth_AP/README.md [hunk 17]
```diff
@@ -0,0 +1,7 @@
+# Smooth_AP
+
+Referenced from the ECCV '20 paper ["Smooth-AP: Smoothing the Path Towards Large-Scale Image Retrieval"](https://www.robots.ox.ac.uk/~vgg/research/smooth-ap/), code is from https://github.com/Andrew-Brown1/Smooth_AP.
+
+The PyTorch implementation of the Smooth-AP loss function is found in src/Smooth_AP_loss.py
```
Comment (sneakerkg) [inline examples/pytorch/hilander/PSS/Smooth_AP/README.md:5]: Do we still need this?

**Comment by sneakerkg** [comment]
@mufeili @jermainewang , as discussed, we will first check-in the PSS model into dgl example to meet the ECCV camera ready requirement (code should be released as promised). The code has been put into the HiLANDER example folder as it is an application of HiLANDER. Duplicated code has been removed. 
Meanwhile, we are applying open-sourcing in [aws-research](https://github.com/amazon-research), when that is approved. We will migrate the code there and left a README redirecting to the new repo.
Let me know if you have further concerns.
